### PR TITLE
Numpy >= 1.13, matplotlib >=2.0 for astropy 3.1

### LINKED ISSUE #7545: Remove numpy hack in io.fits when Numpy minversion is 1.12
I stumbled upon `io/fits/_numpy_hacks.py`, which has a comment that says it can be retired when numpy/numpy#6430 is merged. That Numpy fix was incorporated into Numpy 1.12, which means we can remove this hack when our `minversion` for Numpy is 1.12.

(I am tempted to set this to novice but I probably shouldn't.)

### PR DESCRIPTION
**Pull request by mhvk**
As discussed on the mailing list, for 3.1 we may want to depend on numpy >=1.13, since this allows much better performance for `Quantity` (and thus coordinates, etc.) because it introduced `__array_ufunc__`. This PR removes the stuff needed for numpy < 1.13.

Note that with this PR, 3.1 will likely only be compatible with numpy 1.13 and 1.14, so 2 versions, though 1.15 would presumably be around the corner by then.

fixes #7545

### CODE REVIEW

**Comment by bsipocz** [comment]
(a minor note for the above, we can obviously make most efforts to be np 1.15 compatible even if that release is not out, and add any necessary fixes for full compatibility in a 3.1.x bugfix)

File: CHANGES.rst [hunk 1]
```diff
@@ -215,8 +215,7 @@ astropy.wcs
 Other Changes and Additions
 ---------------------------
 
-- Nothing changed yet.
-
+- Versions of Numpy <1.10 are no longer supported. [#7058]
```
Comment (astrofrog) [inline CHANGES.rst]: Do you mean <1.13?

**Comment by mhvk** [comment]
OK, thanks, now fixed.  I don't know how to change the `circleci` setup - I think we need a new image? Also, the matplotlib 1.5 failure is odd. It brings up a question of whether we should still support that, or make 3.1 a fairly clean break with old cruft. (E.g., could go for scipy >= 1.0 as well...)

**Comment by astrofrog** [comment]
@mhvk - can you rebase? We now have an updated CircleCI configuration, so hopefully this will be fixed.

**Comment by mhvk** [comment]
@astrofrog - OK, rebased. Hopefully tests still pass!

**Comment by astrofrog** [comment]
@mhvk - a few small comments:

* Can you change ``astropy/image-tests-py35-mpl153:1.0`` to ``astropy/image-tests-py35-mpl153:1.1`` in ``.circleci/config.yml``? (I updated Numpy in there). Also, 

* If I try and run ``python setup.py install``, I get the following rather obscure message:

```
NotImplementedError: __array_wrap__ should not be used with a context any more, since we require numpy >=1.13.  Please raise an issue on https://github.com/astropy/astropy
```

Can you move:

```python
if not _ASTROPY_SETUP_:
    _check_numpy()
```

sooner in ``astropy/__init__.py``?

* For the failing Travis build, I think there's a chance this could be real and is due to remaining open files. I wonder if something has changed with Numpy memory mapping. Pinging @drdavella.

**Comment by mhvk** [comment]
@astrofrog - I made most of the smaller changes, but do not understand where you would like me to move the numpy check in `__init__.py` - it already is just after the required part that defines whether or not we're in setup.

**Comment by mhvk** [comment]
@astrofrog - mpl153 is failing. Do you have any idea? It does seem related to something in numpy.

File: .circleci/config.yml [hunk 2]
```diff
@@ -14,7 +14,7 @@ jobs:
           command: PYTHONHASHSEED=42 /opt/python/cp36-cp36m/bin/python setup.py test --parallel=4
   image-tests-mpl153:
     docker:
-      - image: astropy/image-tests-py35-mpl153:1.0
+      - image: astropy/image-tests-py35-mpl153:1.1
```
Comment (astrofrog) [inline .circleci/config.yml]: @mhvk - can you try changing to ``1.2``? I think that should hopefully fix it.

Comment (mhvk) [reply .circleci/config.yml]: OK, changed (merged into last commit).

**Comment by mhvk** [comment]
@astrofrog - the 153 circle-io file still seems to use numpy < 1.13...  I think that failure also explains your earlier comment that it would be nice if the error message was clearer -- somehow `get_package_info()` seems to require importing more than perhaps it should? I must admit being somewhat at a loss...

**Comment by astrofrog** [comment]
I can investigate this during the week of March 19th

**Comment by mhvk** [comment]
Rebased, in part to see if the mpl153 got magically resolved... (though I still think it is fine to start requiring matplotlib >=2.0 for astropy 3.1)

**Comment by astrofrog** [comment]
I agree about requiring Matplotlib >= 2.0, as 1.5 is reasonably ancient now. The issue is indeed that the docker image uses Numpy 1.10

**Comment by mhvk** [comment]
OK, I removed the offending circle-ci build, and also changed the travis test that used matplotlib 1.5.

**Comment by astrofrog** [comment]
How should we enforce the minimum version for Matplotlib - should we raise an error when trying to import astropy.visualization?

**Comment by mhvk** [comment]
@astrofrog - I thought the circle-ci failure was due to that image including numpy < 1.13, not directly because of matplotlib. So, the only decision we are making here is not to *test* against older versions, but I think they should still work. I think this is similar to what we do for scipy. That said, I should add a note to that extent in the changelog...

**Comment by mhvk** [comment]
I'm completely at a loss for the travis failure using "pytest" directly, where somehow an `erf` function ends up among the numpy ufuncs. I cannot reproduce this locally. Now running one with standard python 3.6 and numpy 1.14 to see whether that makes a difference...

**Comment by mhvk** [comment]
The problem is independent of version: when running with pytest, `numpy.core.umath.__dict__` apparently contains an `erf` ufunc, which really shouldn't be there at all.

Right now, my tendency would be to rip out the `pytest` test, but I welcome better solutions!

**Comment by mhvk** [comment]
Rebased with a last commit that should not be merged, but is just to try to diagnose the problem that, somehow, `np.core.umath` gets polluted with `erf` - will that be something else if we do not reference `erf` at all in our code?

**Comment by mhvk** [comment]
OK, it cannot have anything to do with a left-over of a previous install of our code or so -- it is still `erf` that is there when we don't have `erf`. Could there be something fishy with the numpy version installed?

**Comment by mhvk** [comment]
OK, I went ahead and just put an `xfail` in if `erf` is present. Very ugly, but I don't think this should be held up by a test that is not that important. The only risk now would be if numpy does start to include `erf`...

**Comment by astrofrog** [comment]
@mhvk -  locally I see an erf function in Numpy:

```
In [1]: import numpy as np

In [2]: np.__version__
Out[2]: '1.13.3'

In [3]: np.core.umath.erf
Out[3]: <ufunc 'erf'>
```

Could this be platform-specific? (I'm using MacOS X, with Numpy from conda)

**Comment by saimn** [comment]
I have it on a Linux/Conda environment, but not on a Pyenv/Pip one, so probably a difference between OpenBlas and the MKL ?

**Comment by mhvk** [comment]
OK, I think it is an addition that is not upstream at all. But since it is obvious what to do with `erf`, I'll just check in `quantity_helpers` whether it is there and add it if so.

**Comment by mhvk** [comment]
OK, tests now all pass. @astrofrog - you had some request above (https://github.com/astropy/astropy/pull/7058#issuecomment-364136565) above moving `_check_numpy()` further up - I didn't understand exactly where it could go.

**Comment by pllim** [comment]
Any chance #7545 can be addressed here as well?

**Comment by mhvk** [comment]
@pllim - Done with latest commit! (Took a bit of looking at the code to even understand what it meant to do...)

**Comment by saimn** [comment]
@mhvk - It seems that you forgot to remove the imports:
```
  File "/tmp/astropy-test-vw72f7xy/lib/python3.6/site-packages/astropy/io/fits/column.py", line 18, in <module>
    from . import _numpy_hacks as nh
ImportError: cannot import name '_numpy_hacks'
```
and I guess there is the same for `fits/hdu/table.py`.

I also had a look at this code, it's not clear what it does and if it is also needed for the multi-dimensional columns (TDIM). Let's see if tests pass :)

**Comment by mhvk** [comment]
@saimn - ai, that's what one gets for trying to do things too quickly. Don't quite understand how tests passed locally, but probably a matter of not cleaning out `build` properly. Hopefully it now works.

File: astropy/__init__.py [hunk 3]
```diff
@@ -11,7 +11,8 @@
 import os
 from warnings import warn
 
-__minimum_numpy_version__ = '1.10.0'
+__minimum_python_version__ = '3.5'
+__minimum_numpy_version__ = '1.13.0'
```
Comment (saimn) [inline astropy/__init__.py]: `__minimum_python_version__` should not appear here ? (maybe a rebase conflict ?)

Comment (mhvk) [reply astropy/__init__.py]: Thanks - I think I intended to move it up here, but seem to have forgotten to delete the definition lower down.

**Comment by saimn** [comment]
@mhvk - One the build is failing, interesting (why only this one ?) :
```
_________________________ TestUtils.test_realign_dtype _________________________
self = <astropy.io.fits.tests.test_util.TestUtils object at 0x7feeba985588>
    def test_realign_dtype(self):
        """
            Tests a few corner-cases for numpy dtype creation.
    
            These originally were the reason for having a realign_dtype hack.
            """
    
        dt = np.dtype([('a', np.int32), ('b', np.int16)])
        names = dt.names
        formats = tuple(f[0] for f in dt.fields.values())
        dt2 = np.dtype({'names': names, 'formats': formats,
                        'offsets': [0, 0]})
        assert dt2.itemsize == 4
    
        dt2 = np.dtype({'names': names, 'formats': formats,
                        'offsets': [0, 1]})
>       assert dt2.itemsize == 4
E       AssertionError: assert 5 == 4
E        +  where 5 = dtype({'names':['a','b'], 'formats':['<i2','<i4'], 'offsets':[0,1], 'itemsize':5}).itemsize
```

**Comment by mhvk** [comment]
@saimn - very strange. The only thing I noticed about it, is that it still uses matplotlib 1.5, which should not be there. Running it again without that, just in case it makes a difference.

**Comment by mhvk** [comment]
@saimn - it was a matter of assuming dict-like `.values()` have guaranteed order...

**Comment by mhvk** [comment]
The coverage failure is, I think, just because quite a number of lines got removed.

@astrofrog - anything remaining here?

**Comment by bsipocz** [comment]
@mhvk - I agree to ignore coveralls, it complains about dropped coverage in generated c code, but then there is no diff for those, so difficult to do anything really.

**Comment by mhvk** [comment]
@astrofrog, are you happy with this as is?

**Comment by mhvk** [comment]
@astrofrog - I looked again at this move to numpy 1.13, and do think it is all OK. Could you have a look or dismiss your review and ask someone else?

**Comment by pllim** [comment]
I'm merging this now to simplify work of integrating with Python 3.7. Thanks!

**Pull request merged**